runtime/src/utils/buffer/write.rs

```diff
@@ -53,64 +132,138 @@ impl<B: Blob> Write<B> {
-        if buf_len > self.capacity {
-            self.blob.write_at(buf, self.position).await?;
-            self.position += buf_len as u64;
-        } else {
-            self.buffer.extend_from_slice(buf.as_ref());
+        // Write operation can be merged into the existing buffer if:
+        // a) Write starts at or after the buffer's starting position in the blob.
+        // b) The end of the write, relative to the buffer's start, fits within buffer's capacity.
+        let can_write_into_buffer = offset >= buffer_start
+            && (offset - buffer_start) + data_len as u64 <= inner.capacity as u64;
+        if can_write_into_buffer {
+            let buffer_internal_offset = (offset - buffer_start) as usize;
+            let required_buffer_len = buffer_internal_offset + data_len;
+            if required_buffer_len > inner.buffer.len() {
+                inner.buffer.resize(required_buffer_len, 0u8);
+            }
+            inner.buffer[buffer_internal_offset..required_buffer_len].copy_from_slice(data);
+            return Ok(());
+        }
+
+        // All other cases (e.g., write is before buffer, straddles, or would overflow capacity)
+        if !inner.buffer.is_empty() {
+            inner.flush().await?;
         }
+        inner.blob.write_at(buf, offset).await?;
+        inner.position = offset + data_len as u64;
         Ok(())
     }
 
-    /// Flushes buffered data to the underlying blob. Does nothing if the buffer is empty.
-    ///
-    /// If the write to the underlying blob fails, the buffer will be reset (and any pending data not yet
-    /// written will be lost).
-    async fn flush(&mut self) -> Result<(), Error> {
-        if self.buffer.is_empty() {
-            return Ok(());
+    async fn truncate(&self, len: u64) -> Result<(), Error> {
+        // Acquire a write lock on the inner state
+        let mut inner = self.inner.write().await;
+
+        // Prepare the buffer boundaries
+        let buffer_start = inner.position;
+        let buffer_len = inner.buffer.len() as u64;
+        let buffer_end = buffer_start + buffer_len;
+
+        // Adjust buffer content based on `len`
+        if len <= buffer_start {
+            // Truncation point is before or exactly at the start of the buffer.
+            //
+            // All buffered data is now invalid/beyond the new length.
+            inner.buffer.clear();
+            inner.blob.truncate(len).await?;
+            inner.position = len;
+        } else if len < buffer_end {
+            // Truncation point is within the buffer.
+            //
+            // `len` is > `buffer_start_blob_offset` here.
+            // New length of data *within the buffer* is `len - buffer_start_blob_offset`.
+            let new_buffer_actual_len = (len - buffer_start) as usize;
+            inner.buffer.truncate(new_buffer_actual_len);
+        } else {
+            // Truncation point is at or after the end of the buffer, so no changes are needed
```

Comment:
nit: should we remove this empty else? It's a no-op but idk if we do it elsewhere.

Reply:
I believe I've done this elsewhere (I suspect the compiler optimizes it out anyways).